File: scripts/capital-export.py
#!/usr/bin/env python3
"""Backup workbook from the shared book. Never writes the 17 Aug original."""
from __future__ import annotations

import json
import os
import shutil
import urllib.request
from datetime import datetime, timezone
from pathlib import Path

import openpyxl

ROOT = Path(__file__).resolve().parents[1]
ORIG = Path.home() / "Developer/Forge-Capital/260817 Master Investor Tracker TF (CANONICAL).xlsx"
ENV = ROOT / ".env.local"

def load_env() -> dict[str, str]:
    env: dict[str, str] = {}
    if not ENV.exists():
        return env
    for line in ENV.read_text().splitlines():
        line = line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        k, v = line.split("=", 1)
        env[k.strip()] = v.strip().strip('"').strip("'")
    return env


def rest_all(url: str, key: str, schema: str, table: str, select: str) -> list:
    out = []
    start = 0
    while True:
        req = urllib.request.Request(
            f"{url}/rest/v1/{table}?select={select}&offset={start}&limit=1000"
        )
        req.add_header("apikey", key)
        req.add_header("Authorization", f"Bearer {key}")
        req.add_header("Accept-Profile", schema)
        req.add_header("Range", f"{start}-{start+999}")
        with urllib.request.urlopen(req, timeout=60) as resp:
            batch = json.loads(resp.read().decode())
        out.extend(batch)
        if len(batch) < 1000:
            break
        start += 1000
    return out


env = load_env()
url, key = env.get("FORGE_CAPITAL_DB_URL"), env.get("FORGE_CAPITAL_DB_SERVICE_ROLE")
if not url or not key:
    raise SystemExit("FORGE_CAPITAL_DB_* missing")
if not ORIG.exists():
    raise SystemExit(f"template missing {ORIG}")

stamp_before = ORIG.stat().st_mtime
firms = rest_all(url, key, "core", "firms", "id,canonical_name,website_domain,sectors,dnc,notes")
people = rest_all(
    url, key, "core", "people",
    "id,firm_id,full_name,email,email_state,dnc,linkedin_url,role_title,provenance",
)
parts = rest_all(
    url, key, "engage", "participations",
    "firm_id,person_id,mandate_id,stage,status_note,first_sent,latest_touch",
)
mandates = rest_all(url, key, "engage", "mandates", "id,code")
code_by = {m["id"]: m["code"] for m in mandates}
people_by: dict[str, list] = {}
for p in people:
    people_by.setdefault(p["firm_id"], []).append(p)
parts_by: dict[str, list] = {}
for p in parts:
    parts_by.setdefault(p["firm_id"], []).append(p)

yymmdd = datetime.now(timezone.utc).strftime("%y%m%d")
out = Path.home() / f"Developer/Forge-Capital/{yymmdd} Master Investor Tracker TF (CANONICAL).xlsx"
if out.resolve() == ORIG.resolve() or "260817" in out.name:
    raise SystemExit("refusing to write the 17 Aug original")
shutil.copy2(ORIG, out)

wb = openpyxl.load_workbook(out)
ws = wb["Master Tracker"]
# Keep header rows 1-2. Clear old data from row 3 down.
if ws.max_row > 2:
    ws.delete_rows(3, ws.max_row - 2)

CODES = ["SK", "FF", "PA", "SS", "CA", "US", "OD", "HO"]
# 1-based Excel columns: Latest in each block
LATEST = [17, 22, 27, 32, 37, 42, 47, 52]
DAYS = [18, 23, 28, 33, 38, 43, 48, 53]

sorted_firms = sorted(firms, key=lambda f: (f.get("canonical_name") or "").lower())
for i, f in enumerate(sorted_firms):
    r = i + 3
    plist = people_by.get(f["id"]) or []
    named = next((p for p in plist if not p.get("dnc")), plist[0] if plist else None)
    pp = parts_by.get(f["id"]) or []
    by_code = {}
    for p in pp:
        c = code_by.get(p["mandate_id"])
        if c:
            by_code[c] = p
    ws.cell(r, 10, f.get("canonical_name"))
    ws.cell(r, 11, f.get("website_domain"))
    ws.cell(r, 12, named.get("full_name") if named else None)
    ws.cell(r, 13, named.get("email") if named else None)
    sectors = f.get("sectors")
    sector = ", ".join(sectors) if isinstance(sectors, list) else sectors
    if f.get("dnc"):
        sector = " · ".join(x for x in [sector, "DNC"] if x)
    ws.cell(r, 14, sector)
    ws.cell(r, 15, len(pp) or None)
    max_bits = []
    for j, code in enumerate(CODES):
        p = by_code.get(code)
        tick_col = 2 + j
        first_col = 16 + j * 5
        latest_col = first_col + 1
        days_col = first_col + 2
        status_col = first_col + 3
        comm_col = first_col + 4
        latest_letter = openpyxl.utils.get_column_letter(latest_col)
        max_bits.append(f"{latest_letter}{r}")
        if not p:
            ws.cell(r, days_col, f'=IF({latest_letter}{r}="","",TODAY()-{latest_letter}{r})')
            continue
        ws.cell(r, tick_col, "✓")
        first = (p.get("first_sent") or "")[:10] or None
        latest = (p.get("latest_touch") or "")[:10] or None
        ws.cell(r, first_col, first)
        ws.cell(r, latest_col, latest)
        ws.cell(r, days_col, f'=IF({latest_letter}{r}="","",TODAY()-{latest_letter}{r})')
        ws.cell(r, status_col, p.get("status_note") or p.get("stage"))
        ws.cell(r, comm_col, p.get("status_note"))
    mx = ",".join(max_bits)
    ws.cell(r, 1, f'=IF(MAX({mx})=0,"",TODAY()-MAX({mx}))')

def clear_sheet(ws, header_rows: int) -> None:
    if ws.max_row > header_rows:
        ws.delete_rows(header_rows + 1, ws.max_row - header_rows)


def write_rows(ws, start_row: int, rows: list[list]) -> None:
    for i, row in enumerate(rows):
        for c, val in enumerate(row, 1):
            ws.cell(start_row + i, c, val)


firms_by = {f["id"]: f for f in firms}
people_by_id = {p["id"]: p for p in people}
od_id = next((m["id"] for m in mandates if m.get("code") == "OD"), None)
us_id = next((m["id"] for m in mandates if m.get("code") == "US"), None)
od_parts = [p for p in parts if p.get("mandate_id") == od_id]
us_parts = [p for p in parts if p.get("mandate_id") == us_id]


def firm_name(fid):
    f = firms_by.get(fid) or {}
    return f.get("canonical_name")


def firm_web(fid):
    f = firms_by.get(fid) or {}
    return f.get("website_domain")


def person_name(pid):
    p = people_by_id.get(pid) or {}
    return p.get("full_name")


def person_email(pid):
    p = people_by_id.get(pid) or {}
    return p.get("email")


def od_row(p):
    return [
        firm_name(p.get("firm_id")),
        firm_web(p.get("firm_id")),
        p.get("stage"),
        (p.get("latest_touch") or p.get("first_sent") or "")[:10] or None,
        p.get("status_note"),
        person_name(p.get("person_id")),
        person_email(p.get("person_id")),
        p.get("status_note"),
    ]


if "Dashboard" in wb.sheetnames:
    d = wb["Dashboard"]
    d["A1"] = "Shared-book backup — generated from Corpus"
    d["A2"] = datetime.now(timezone.utc).isoformat()
    d["A3"] = f"Firms {len(firms)} · people {len(people)} · participations {len(parts)}"
    d["A4"] = "The 17 Aug original was not modified."

if "README" in wb.sheetnames:
    r = wb["README"]
    r["A1"] = "Generated from the shared book. Excel is a download only."
    r["A2"] = "Do not type in the 17 Aug CANONICAL file."

if "US Exploration (Forge)" in wb.sheetnames:
    us = wb["US Exploration (Forge)"]
    clear_sheet(us, 2)
    write_rows(
        us,
        3,
        [
            [
                firm_name(p.get("firm_id")),
                person_name(p.get("person_id")),
                person_email(p.get("person_id")),
                "book",
                p.get("stage"),
                "",
                p.get("status_note"),
            ]
            for p in us_parts
        ],
    )

if "Odysseus — Jordan full" in wb.sheetnames:
    j = wb["Odysseus — Jordan full"]
    clear_sheet(j, 1)
    write_rows(
        j,
        2,
        [
            [
                firm_name(p.get("firm_id")),
                firm_web(p.get("firm_id")),
                "",
                p.get("stage"),
                p.get("status_note"),
                firm_name(p.get("firm_id")),
                "book",
                person_name(p.get("person_id")),
                person_email(p.get("person_id")),
                "",
                "",
                p.get("status_note"),
            ]
            for p in od_parts
        ],
    )

do_not = [
    p for p in od_parts
    if (p.get("stage") in ("disqualified", "blocked", "closed_lost"))
    or "do not" in (p.get("status_note") or "").lower()
    or "cautious" in (p.get("status_note") or "").lower()
    and "leave" in (p.get("status_note") or "").lower()
]
low = [p for p in od_parts if "cautious" in (p.get("status_note") or "").lower()]
cands = [p for p in od_parts if (p.get("stage") or "") in ("research", "awaiting_signoff")]

if "Odysseus — DO NOT outreach" in wb.sheetnames:
    ws = wb["Odysseus — DO NOT outreach"]
    clear_sheet(ws, 1)
    write_rows(ws, 2, [od_row(p) for p in do_not])

if "Odysseus — Low priority OK" in wb.sheetnames:
    ws = wb["Odysseus — Low priority OK"]
    clear_sheet(ws, 1)
    write_rows(ws, 2, [od_row(p) for p in low])

if "Odysseus — Candidates approval" in wb.sheetnames:
    ws = wb["Odysseus — Candidates approval"]
    clear_sheet(ws, 1)
    write_rows(
        ws,
        2,
        [
            [
                firm_name(p.get("firm_id")),
                firm_web(p.get("firm_id")),
                person_name(p.get("person_id")),
                "",
                "",
                "",
                "",
                "",
                "",
                p.get("status_note"),
            ]
            for p in cands
        ],
    )

if "Odysseus — Jordan DO-NOT-APPROA" in wb.sheetnames:
    ws = wb["Odysseus — Jordan DO-NOT-APPROA"]
    clear_sheet(ws, 1)
    write_rows(
        ws,
        2,
        [
            [
                firm_name(p.get("firm_id")),
                firm_web(p.get("firm_id")),
                person_name(p.get("person_id")),
                person_email(p.get("person_id")),
                p.get("stage"),
                firm_web(p.get("firm_id")),
            ]
            for p in do_not
        ],
    )

li = [p for p in people if (p.get("provenance") or "").lower() == "linkedin"]
if "LinkedIn Connections" in wb.sheetnames:
    ws = wb["LinkedIn Connections"]
    clear_sheet(ws, 1)
    write_rows(
        ws,
        2,
        [
            [
                p.get("full_name"),
                firm_name(p.get("firm_id")),
                p.get("role_title"),
                p.get("linkedin_url"),
                "",
                p.get("email"),
                "yes" if p.get("firm_id") in parts_by else "no",
            ]
            for p in li
        ],
    )

if "Generated" not in wb.sheetnames:
    g = wb.create_sheet("Generated")
    g["A1"] = "Generated from the shared book. Do not type in this file."
    g["A2"] = "The 17 Aug original was not modified."
    g["A3"] = datetime.now(timezone.utc).isoformat()
    g["A4"] = f"Firms {len(firms)} · people {len(people)} · participations {len(parts)}"

wb.save(out)
wb.close()
if ORIG.stat().st_mtime != stamp_before:
    raise SystemExit("ERROR: original workbook mtime changed")
print(json.dumps({
    "out": str(out),
    "firms": len(firms),
    "people": len(people),
    "participations": len(parts),
    "data_rows": len(sorted_firms),
    "orig_untouched": True,
}))

File: scripts/capital-export.mjs
#!/usr/bin/env node
/**
 * Backup workbook FROM the shared book. Never writes the 260817 original.
 * Default output: ~/Developer/Forge-Capital/YYMMDD Master Investor Tracker TF (CANONICAL).xlsx
 */
import { existsSync, readFileSync, writeFileSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { homedir } from "node:os";
import { createRequire } from "node:module";
import { createClient } from "@supabase/supabase-js";

const require = createRequire(import.meta.url);
const XLSX = require("xlsx");
const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const FORBIDDEN = resolve(
  homedir(),
  "Developer/Forge-Capital/260817 Master Investor Tracker TF (CANONICAL).xlsx",
);

function loadEnv() {
  const p = resolve(ROOT, ".env.local");
  if (!existsSync(p)) return;
  for (const raw of readFileSync(p, "utf8").split("\n")) {
    const line = raw.trim();
    if (!line || line.startsWith("#") || !line.includes("=")) continue;
    const i = line.indexOf("=");
    let v = line.slice(i + 1).trim();
    if ((v.startsWith('"') && v.endsWith('"')) || (v.startsWith("'") && v.endsWith("'")))
      v = v.slice(1, -1);
    process.env[line.slice(0, i).trim()] = v;
  }
}
loadEnv();

const url = process.env.FORGE_CAPITAL_DB_URL;
const key = process.env.FORGE_CAPITAL_DB_SERVICE_ROLE;
if (!url || !key) {
  console.error("FORGE_CAPITAL_DB_* missing");
  process.exit(2);
}
const opts = { auth: { persistSession: false, autoRefreshToken: false } };
const core = createClient(url, key, { ...opts, db: { schema: "core" } });
const engage = createClient(url, key, { ...opts, db: { schema: "engage" } });

async function allRows(client, table, select) {
  const out = [];
  for (let from = 0; from < 50000; from += 1000) {
    const { data, error } = await client.from(table).select(select).range(from, from + 999);
    if (error) throw new Error(table + " " + error.message);
    out.push(...(data ?? []));
    if (!data || data.length < 1000) break;
  }
  return out;
}

const [firms, people, parts, mandates] = await Promise.all([
  allRows(core, "firms", "id, canonical_name, website_domain, sectors, dnc, notes"),
  allRows(core, "people", "id, firm_id, full_name, email, email_state, dnc"),
  allRows(engage, "participations", "firm_id, person_id, mandate_id, stage, status_note, first_sent, latest_touch"),
  engage.from("mandates").select("id, code").then((r) => {
    if (r.error) throw new Error(r.error.message);
    return r.data ?? [];
  }),
]);

const codeByMandate = Object.fromEntries(mandates.map((m) => [m.id, m.code]));
const peopleByFirm = new Map();
for (const p of people) {
  if (!peopleByFirm.has(p.firm_id)) peopleByFirm.set(p.firm_id, []);
  peopleByFirm.get(p.firm_id).push(p);
}
const partsByFirm = new Map();
for (const p of parts) {
  if (!partsByFirm.has(p.firm_id)) partsByFirm.set(p.firm_id, []);
  partsByFirm.get(p.firm_id).push(p);
}

const TEMPLATE = resolve(
  homedir(),
  "Developer/Forge-Capital/260817 Master Investor Tracker TF (CANONICAL).xlsx",
);
const CODES = ["SK", "FF", "PA", "SS", "CA", "US", "OD", "HO"];
const LATEST_COLS = ["Q", "V", "AA", "AF", "AK", "AP", "AU", "AZ"]; // 1-based Excel letters for Latest in each block
const header1 = [
  "LAST CONTACT", "CONTACTED?", null, null, null, null, null, null, null,
  "INVESTOR INFO", null, null, null, null, null,
  "SKYSAILS POWER", null, null, null, null,
  "FISHFROM", null, null, null, null,
  "PANATERE", null, null, null, null,
  "SPACE SOLAR", null, null, null, null,
  "CASPER FUNDING", null, null, null, null,
  "US ARBITRAGE", null, null, null, null,
  "ODYSSEUS SPACE", null, null, null, null,
  "HOOLEY RF", null, null, null, null,
];
const header2 = [
  "Days ago", "SkySails", "FishFrom", "Panatere", "Space Solar", "Casper", "US Arb", "Odysseus", "Hooley",
  "Investor", "Website", "Contact", "Email", "Sector", "# Co",
];
for (let i = 0; i < 8; i++) header2.push("First Sent", "Latest", "Days Since", "Status", "Commentary");

const aoa = [header1, header2];
const sorted = firms.slice().sort((a, b) =>
  String(a.canonical_name).localeCompare(String(b.canonical_name), "en"),
);

for (const f of sorted) {
  const plist = peopleByFirm.get(f.id) ?? [];
  const named = plist.find((p) => !p.dnc) ?? plist[0];
  const pp = partsByFirm.get(f.id) ?? [];
  const byCode = {};
  for (const p of pp) {
    const c = codeByMandate[p.mandate_id];
    if (c) byCode[c] = p;
  }
  const row = new Array(55).fill(null);
  row[9] = f.canonical_name;
  row[10] = f.website_domain;
  row[11] = named?.full_name ?? null;
  row[12] = named?.email ?? null;
  row[13] = Array.isArray(f.sectors) ? f.sectors.join(", ") : f.sectors;
  row[14] = pp.length || null;
  if (f.dnc) row[13] = [row[13], "DNC"].filter(Boolean).join(" | ");
  CODES.forEach((code, i) => {
    const p = byCode[code];
    const base = 15 + i * 5;
    if (!p) return;
    row[1 + i] = "✓";
    row[base] = p.first_sent ? String(p.first_sent).slice(0, 10) : null;
    row[base + 1] = p.latest_touch ? String(p.latest_touch).slice(0, 10) : null;
    row[base + 3] = p.status_note || p.stage;
    row[base + 4] = p.status_note;
  });
  aoa.push(row);
}

if (!existsSync(TEMPLATE)) {
  console.error("template missing", TEMPLATE);
  process.exit(2);
}
const template = XLSX.readFile(TEMPLATE, { cellDates: true, cellFormula: true });
const master = XLSX.utils.aoa_to_sheet(aoa);
const DAYS_COLS = ["R", "W", "AB", "AG", "AL", "AQ", "AV", "BA"];
const lastRow = 2 + sorted.length;
for (let r = 3; r <= lastRow; r += 1) {
  const maxList = LATEST_COLS.map((c) => `${c}${r}`).join(",");
  master[`A${r}`] = {
    t: "n",
    f: `IF(MAX(${maxList})=0,"",TODAY()-MAX(${maxList}))`,
  };
  DAYS_COLS.forEach((daysCol, i) => {
    const latest = LATEST_COLS[i];
    master[`${daysCol}${r}`] = {
      t: "n",
      f: `IF(${latest}${r}="","",TODAY()-${latest}${r})`,
    };
  });
}
if (master["!ref"]) {
  master["!ref"] = XLSX.utils.encode_range({
    s: { r: 0, c: 0 },
    e: { r: lastRow - 1, c: 54 },
  });
}
const wb = {
  SheetNames: [...template.SheetNames],
  Sheets: {},
};
for (const name of template.SheetNames) {
  wb.Sheets[name] = name === "Master Tracker" ? master : template.Sheets[name];
}
if (!wb.SheetNames.includes("Generated")) wb.SheetNames.push("Generated");
wb.Sheets.Generated = XLSX.utils.aoa_to_sheet([
  ["Generated from the shared book (core/engage). Do not type in this file."],
  ["The 17 Aug original was not modified."],
  ["Generated", new Date().toISOString()],
  ["Firms", firms.length],
  ["People", people.length],
  ["Participations", parts.length],
]);

const yymmdd = new Date().toISOString().slice(2, 10).replace(/-/g, "");
const out = resolve(
  homedir(),
  `Developer/Forge-Capital/${yymmdd} Master Investor Tracker TF (CANONICAL).xlsx`,
);
if (resolve(out) === resolve(FORBIDDEN) || out.includes("260817")) {
  console.error("refusing to write the 17 Aug original");
  process.exit(2);
}
writeFileSync(out, XLSX.write(wb, { type: "buffer", bookType: "xlsx" }));
console.log({ out, firms: firms.length, people: people.length, participations: parts.length, data_rows: aoa.length - 2 });
